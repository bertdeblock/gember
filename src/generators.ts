import { generate } from "./generate.js";

export function generateComponent(
  name: string,
  packagePath: string,
  {
    classBased = false,
    nested,
    path,
    typescript = false,
  }: {
    classBased?: boolean;
    nested?: boolean;
    path?: string;
    typescript?: boolean;
  } = {},
): Promise<void> {
  return generate({
    customTargetPath: path,
    entityName: name,
    generatorName: "component",
    nested,
    packagePath,
    targetDir: "components",
    templateFilename:
      (classBased ? "class-based" : "template-only") +
      (typescript ? ".gts" : ".gjs"),
  });
}

export function generateHelper(
  name: string,
  packagePath: string,
  {
    classBased = false,
    path,
    typescript = false,
  }: {
    classBased?: boolean;
    path?: string;
    typescript?: boolean;
  } = {},
): Promise<void> {
  return generate({
    customTargetPath: path,
    entityName: name,
    generatorName: "helper",
    packagePath,
    targetDir: "helpers",
    templateFilename:
      (classBased ? "class-based" : "function-based") +
      (typescript ? ".ts" : ".js"),
  });
}

export function generateModifier(
  name: string,
  packagePath: string,
  {
    classBased = false,
    path,
    typescript = false,
  }: {
    classBased?: boolean;
    path?: string;
    typescript?: boolean;
  } = {},
): Promise<void> {
  return generate({
    customTargetPath: path,
    entityName: name,
    generatorName: "modifier",
    packagePath,
    targetDir: "modifiers",
    templateFilename:
      (classBased ? "class-based" : "function-based") +
      (typescript ? ".ts" : ".js"),
  });
}

export function generateService(
  name: string,
  packagePath: string,
  {
    path,
    typescript = false,
  }: {
    path?: string;
    typescript?: boolean;
  } = {},
): Promise<void> {
  return generate({
    customTargetPath: path,
    entityName: name,
    generatorName: "service",
    packagePath,
    targetDir: "services",
    templateFilename: typescript ? "service.ts" : "service.js",
  });
}
