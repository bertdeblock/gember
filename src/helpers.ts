import type { EmberPackageJson } from "./types.js";

export function isV1Addon(packageJson: EmberPackageJson): boolean {
  if (isAddon(packageJson)) {
    const { version } = packageJson["ember-addon"] ?? {};

    return version === 1 || version === undefined;
  }

  return false;
}

export function isV2Addon(packageJson: EmberPackageJson): boolean {
  if (isAddon(packageJson)) {
    const { version } = packageJson["ember-addon"] ?? {};

    return version === 2;
  }

  return false;
}

function isAddon(packageJson: EmberPackageJson): boolean {
  if (Array.isArray(packageJson.keywords)) {
    return packageJson.keywords.includes("ember-addon");
  }

  return false;
}
