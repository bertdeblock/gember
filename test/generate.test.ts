import { afterEach, it } from "vitest";
import { generateComponent } from "../src/generators.ts";
import { Package } from "./helpers.ts";

let pkg: Package;

afterEach(() => pkg.cleanUp());

it("supports v1 apps", async (ctx) => {
  pkg = await Package.create("v1-app");

  await generateComponent("foo", pkg.path);

  const content = await pkg.readFile("app/components/foo.gjs");

  ctx.expect(content).toMatchSnapshot();
});

it("supports v2 apps", async (ctx) => {
  pkg = await Package.create("v2-app");

  await generateComponent("foo", pkg.path);

  const content = await pkg.readFile("app/components/foo.gjs");

  ctx.expect(content).toMatchSnapshot();
});

it("supports v1 addons", async (ctx) => {
  pkg = await Package.create("v1-addon");

  await generateComponent("foo", pkg.path);

  const content = await pkg.readFile("addon/components/foo.gjs");

  ctx.expect(content).toMatchSnapshot();
});

it("supports v2 addons", async (ctx) => {
  pkg = await Package.create("v2-addon");

  await generateComponent("foo", pkg.path);

  const content = await pkg.readFile("src/components/foo.gjs");

  ctx.expect(content).toMatchSnapshot();
});
