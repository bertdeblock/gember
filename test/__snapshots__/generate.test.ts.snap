// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`supports v1 addons 1`] = `
"<template>{{yield}}</template>
"
`;

exports[`supports v1 apps 1`] = `
"<template>{{yield}}</template>
"
`;

exports[`supports v2 addons 1`] = `
"<template>{{yield}}</template>
"
`;

exports[`supports v2 apps 1`] = `
"<template>{{yield}}</template>
"
`;
